src/coordinator/ts/series.go

```diff
@@ -49,3 +54,64 @@ func (b *Series) Len() int { return b.vals.Len() }
 
 // Values returns the underlying values interface
 func (b *Series) Values() Values { return b.vals }
+
+// Align adjusts the datapoints to start, end and a fixed interval
+func (b *Series) Align(start, end time.Time, interval time.Duration) (*Series, error) {
+	fixedVals, err := alignValues(b.Values(), start, end, interval)
+	if err != nil {
+		return nil, err
+	}
+
+	return NewSeries(b.name, fixedVals, b.Tags), nil
+}
+
+func alignValues(values Values, start, end time.Time, interval time.Duration) (FixedResolutionMutableValues, error) {
+	switch vals := values.(type) {
+	case Datapoints:
+		return RawPointsToFixedStep(vals, start, end, interval)
+	case FixedResolutionMutableValues:
+		// TODO: Align fixed resolution as well once storages can return those directly
+		return vals, nil
+	default:
+		return nil, fmt.Errorf("unknown type: %v", vals)
+	}
+}
+
+// SeriesList represents a slice of series pointers
+type SeriesList []*Series
+
+// Resolution returns the resolution for a fixed step series list. It assumes all underlying series will have the same resolution
+func (seriesList SeriesList) Resolution() (time.Duration, error) {
+	var resolution time.Duration
+	for i, s := range seriesList {
+		fixedRes, ok := s.Values().(FixedResolutionMutableValues)
+		if !ok {
+			return 0, errors.New("only fixed resolution supported")
+		}
+
+		if i == 0 {
+			resolution = fixedRes.Resolution()
+		} else {
+			if resolution != fixedRes.Resolution() {
+				return 0, fmt.Errorf("resolution mismatch, r1: %v, r2: %v", resolution, fixedRes.Resolution())
+			}
+		}
+	}
+
+	return resolution, nil
+}
+
+// Align aligns each series to the given start, end and step.
+func (seriesList SeriesList) Align(start, end time.Time, interval time.Duration) (SeriesList, error) {
```

Comment:
Seems better to do this inplace on the existing seriesList to reduce allocations

Reply:
It's cleaner this way since if the thing errors out in between then seriesList gets in a strange state. Also, we won't be able to inplace replace Datapoint -> FixedRes values so might not end up saving a lot. Let's see if this becomes a bottleneck ?